File: Sample-Connect-Objective-C-App/ConnectObjective-CSampleApp/ViewController.m
//
//  ViewController.m
//  ConnectObjective-CSampleApp
//
//  Created by Tapas Behera on 8/8/17.
//  Copyright © 2017 Acuant. All rights reserved.
//

#import "ViewController.h"
#import <AcuantMobileSDK/AcuantMobileSDKController.h>
#import "ProgressHUD.h"
#import "ResultViewController.h"

@interface ViewController ()<ResultCancelDelegate,AcuantMobileSDKControllerCapturingDelegate,AcuantMobileSDKControllerProcessingDelegate>

@property(nonatomic,strong) AcuantMobileSDKController* instance;
@property(nonatomic,strong) NSDictionary* resultData;
@property(nonatomic) int side; // 0 front , 1 back
@property(nonatomic) AcuantCardType cardType;

@property(nonatomic,strong) NSString* username;
@property(nonatomic,strong) NSString* password;
@property(nonatomic,strong) NSString* subscription;
@property(nonatomic,strong) NSString* url;

@property(nonatomic,strong) IBOutlet UIImageView* frontCardImageView;
@property(nonatomic,strong) IBOutlet UILabel* frontImageViewLabel;
@property(nonatomic,strong) IBOutlet UILabel* backImageViewLabel;
@property (strong, nonatomic) IBOutlet UIImageView *backCardImageView;
@property (strong, nonatomic) IBOutlet UIButton *processButton;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _username = @"XXXXXXXXXXXX";
    _password = @"XXXXXXXXXXXX";
    _subscription = @"XXXXXXXXXXXX";
    _url = @"https://devconnect.assureid.net/AssureIDService";
    
    [self hideFrontUI];
    [self hideUIBack];
    [self hideProcessButton];
    _side = 0;
    
    //UI Initialization ends
    
    // Create and add the view to the screen.
    
    [ProgressHUD show:@"Validating key"];
    self.view.userInteractionEnabled=NO;
    _instance = [AcuantMobileSDKController initAcuantMobileSDKWithUsername:_username password:_password subscription:_subscription url:_url andDelegate:self];

}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void) showFrontUI{
    _frontCardImageView.layer.cornerRadius = 8.0;
    _frontCardImageView.clipsToBounds = YES;
    _frontCardImageView.hidden=NO;
    _frontCardImageView.userInteractionEnabled=YES;
    _frontCardImageView.image = nil;
    _frontCardImageView.layer.borderWidth=2;
    _frontImageViewLabel.hidden=NO;
}

-(void)hideFrontUI{
    _frontCardImageView.layer.cornerRadius = 8.0;
    _frontCardImageView.clipsToBounds = YES;
    _frontCardImageView.hidden=YES;
    _frontCardImageView.image = nil;
    _frontCardImageView.layer.borderWidth=2;
    _frontImageViewLabel.hidden=YES;
}


-(void) showBackUI{
    _backCardImageView.layer.cornerRadius = 8.0;
    _backCardImageView.clipsToBounds = YES;
    _backCardImageView.hidden=NO;
    _backCardImageView.image = nil;
    _backCardImageView.layer.borderWidth=2;
    _backImageViewLabel.hidden=false;
}
-(void)hideUIBack{
    
    _backCardImageView.layer.cornerRadius = 8.0;
    _backCardImageView.clipsToBounds = YES;
    _backCardImageView.userInteractionEnabled=YES;
    _backCardImageView.hidden=YES;
    _backCardImageView.image = nil;
    _backCardImageView.layer.borderWidth=2;
    _backImageViewLabel.hidden=YES;
    
}
-(void) showProcessButton{
    _processButton.hidden=NO;
    _processButton.enabled=YES;
}

-(void) hideProcessButton{
    _processButton.hidden=YES;
}

-(void) resetData{
    _side = 0;
    _resultData = nil;
    _backCardImageView.image = nil;
    _frontCardImageView.image = nil;
}

// UI Callbacks
-(void) didFinishShowingResult{
    [self resetData];
    [self hideProcessButton];
    [self hideFrontUI];
    [self hideUIBack];
}

- (IBAction)captureTapped:(id)sender {
    [self showFrontUI];
    [self hideUIBack];
    [self resetData];
}
- (IBAction)processTapped:(id)sender {
    self.view.userInteractionEnabled=NO;
    NSString* errorMessage = [self validateState];
    if([errorMessage isEqualToString:@""]){
        [ProgressHUD show:@"Capturing data..."];
        AcuantCardProcessRequestOptions* options  = [AcuantCardProcessRequestOptions defaultRequestOptionsForCardType:_cardType];
        
        [_instance processCardWithOptions:options frontImage:_frontCardImageView.image backImage:_backCardImageView.image barcodeString:nil];
        
        
    }else{
        UIAlertController* alertController = [[UIAlertController alloc] init];
        [alertController setTitle:@"ConnectExample"];
        [alertController setMessage:errorMessage];
        UIAlertAction* dismissButton = [UIAlertAction actionWithTitle:@"OK"
                                                                style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action)
                                        {
                                            [alertController dismissViewControllerAnimated:YES completion:nil];
                                        }];
        
        [alertController addAction:dismissButton];
        [self presentViewController:alertController animated:YES completion:nil];
    }

}

- (IBAction)frontImageTapped:(id)sender {
    _side = 0;
    [self showCamera];
}

- (IBAction)backImageTapped:(id)sender {
    _side = 1;
    [self showCamera];
}

//Show Camera
-(void) showCamera{
    if(_instance!=nil){
        [_instance showManualCameraInterfaceInViewController:self delegate:self cardType:AcuantCardTypeDriversLicenseCard region:AcuantCardRegionUnitedStates andBackSide:YES];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"toResult"]){
        ResultViewController* destinationVC = (ResultViewController*)[segue destinationViewController];
        destinationVC.cancelDelegate = self;
        destinationVC.cardData=_resultData;
        destinationVC.cardType=_cardType;
        destinationVC.username=_username;
        destinationVC.password=_password;
    }
    
}


//Data validation
-(NSString*) validateState{
    NSString* retValue=@"";
    if(_frontCardImageView.image==nil){
        retValue = @"Please provide an ID image";
    }
    return retValue;
}

// Web service callbacks
-(void) mobileSDKWasValidated:(BOOL)wasValidated{
    self.view.userInteractionEnabled=YES;
    dispatch_async(dispatch_get_main_queue(), ^{
        [ProgressHUD dismiss];
    });
    if(!wasValidated){
        
        UIAlertController* alertController = [[UIAlertController alloc] init];
        [alertController setTitle:@"ConnectExample"];
        [alertController setMessage:@"Credntials are not valid"];
        UIAlertAction* dismissButton = [UIAlertAction actionWithTitle:@"OK"
                                                                style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action)
                                        {
                                            [alertController dismissViewControllerAnimated:YES completion:nil];
                                        }];
        
        [alertController addAction:dismissButton];
        [self presentViewController:alertController animated:YES completion:nil];
    }
    
}

-(void) didFinishProcessingCardWithResult:(AcuantCardResult *)result{
    
}

-(void) didFinishProcessingCardWithAssureIDResult:(id)json{
    self.view.userInteractionEnabled=YES;
    _resultData=json;
    dispatch_async(dispatch_get_main_queue(), ^{
        [ProgressHUD dismiss];
        [self performSegueWithIdentifier:@"toResult" sender:nil];
    });
}


// Capture Callbacks

- (BOOL)showBackButton{
    return YES;
}

-(void) didCaptureCropImage:(UIImage *)cardImage scanBackSide:(BOOL)scanBackSide andCardType:(AcuantCardType)cardType{
    _cardType=cardType;
    if(cardType == AcuantCardTypeDriversLicenseCard){
        if(_side==0){
            _frontCardImageView.image=cardImage;
            _frontImageViewLabel.hidden=YES;
            UIAlertController* alertController = [[UIAlertController alloc] init];
            [alertController setTitle:@"ConnectExample"];
            [alertController setMessage:@"Scan the backside of the drivers license"];
            UIAlertAction* dismissButton = [UIAlertAction actionWithTitle:@"OK"
                                                                    style:UIAlertActionStyleDefault
                                                                  handler:^(UIAlertAction * action)
                                            {
                                                [self showCamera];
                                                _side=1;
                                                [alertController dismissViewControllerAnimated:YES completion:nil];
                                            }];
            
            [alertController addAction:dismissButton];
            
            
            [self presentViewController:alertController animated:YES completion:nil];
            
        }else{
            [self showBackUI];
            _backCardImageView.image=cardImage;
            _backImageViewLabel.hidden=YES;
            [self showProcessButton];
        }
    }else{
        _frontCardImageView.image=cardImage;
        _frontImageViewLabel.hidden=YES;
        [self showProcessButton];
    }
    
}

-(void) didCaptureData:(NSString *)data{
    
}


-(void) didFailToCaptureCropImage{
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController* alert = [[UIAlertController alloc] init];
        [alert setTitle:@"ConnectExample"];
        [alert setMessage:@"Failed to capture Image.Please try again"];
        UIAlertAction* dismissButton = [UIAlertAction actionWithTitle:@"OK"
                                                                style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action)
                                        {
                                            [alert dismissViewControllerAnimated:YES completion:nil];
                                        }];
        
        [alert addAction:dismissButton];
        
        [self presentViewController:alert animated:YES completion:nil];
    });
}

-(void) didFinishValidatingImageWithResult:(AcuantCardResult *)result{
    
}


// Errors
-(void) didFailWithError:(AcuantError *)error {
    [ProgressHUD dismiss];
    self.view.userInteractionEnabled=YES;
    [self showErrorAlert:error];
    
}

-(void) showErrorAlert:(AcuantError*)error{
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController* alert = [[UIAlertController alloc] init];
        [alert setTitle:@"ConnectExample"];
        [alert setMessage:error.errorMessage];
        UIAlertAction* dismissButton = [UIAlertAction actionWithTitle:@"OK"
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction * action)
        {
            [alert dismissViewControllerAnimated:YES completion:nil];
        }];
        
        [alert addAction:dismissButton];
        
        [self presentViewController:alert animated:YES completion:nil];
    });

}



@end

File: Sample-Connect-Swift-App/AcuantiOSSDKAssureIDSwiftSample/ViewController.swift
//
//  ViewController.swift
//  AcuantiOSSDKSwiftSample
//
//  Created by Tapas Behera on 7/27/17.
//  Copyright © 2017 Acuant. All rights reserved.
//

import UIKit

protocol ResultCancelDelegate {
    func didFinishShowingResult()
}



class ViewController: UIViewController ,AcuantMobileSDKControllerCapturingDelegate,AcuantMobileSDKControllerProcessingDelegate,ResultCancelDelegate{
    
    
    
    
    struct ScreenSize
    {
        static let SCREEN_WIDTH         = UIScreen.main.bounds.size.width
        static let SCREEN_HEIGHT        = UIScreen.main.bounds.size.height
        static let SCREEN_MAX_LENGTH    = max(ScreenSize.SCREEN_WIDTH, ScreenSize.SCREEN_HEIGHT)
        static let SCREEN_MIN_LENGTH    = min(ScreenSize.SCREEN_WIDTH, ScreenSize.SCREEN_HEIGHT)
    }
    
    struct DeviceType
    {
        static let IS_IPHONE_4_OR_LESS  = UIDevice.current.userInterfaceIdiom == .phone && ScreenSize.SCREEN_MAX_LENGTH < 568.0
        static let IS_IPHONE_5          = UIDevice.current.userInterfaceIdiom == .phone && ScreenSize.SCREEN_MAX_LENGTH == 568.0
        static let IS_IPHONE_6_7          = UIDevice.current.userInterfaceIdiom == .phone && ScreenSize.SCREEN_MAX_LENGTH == 667.0
        static let IS_IPHONE_6P_7P         = UIDevice.current.userInterfaceIdiom == .phone && ScreenSize.SCREEN_MAX_LENGTH == 736.0
        static let IS_IPAD              = UIDevice.current.userInterfaceIdiom == .pad && ScreenSize.SCREEN_MAX_LENGTH == 1024.0
        static let IS_IPAD_PRO          = UIDevice.current.userInterfaceIdiom == .pad && ScreenSize.SCREEN_MAX_LENGTH == 1366.0
    }
    
    var instance:AcuantMobileSDKController!
    var progressHUD:ProgressHUD!
    var resultData:Any!
    var side:Int! // 0 front , 1 back
    var cardType:AcuantCardType!
    
    var username : String = "XXXXXXXXXXXX"
    var password : String = "XXXXXXXXXXXX"
    var subscription : String = "XXXXXXXXXXXX"
    var url : String = "https://devconnect.assureid.net/AssureIDService"
    
    @IBOutlet var backCardImageView: UIImageView!
    @IBOutlet var frontCardImageView: UIImageView!
    @IBOutlet var frontImageViewLabel: UILabel!
    @IBOutlet var backImageViewLabel: UILabel!
    
    @IBOutlet var processButton: UIButton!
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //UI Initialization  starts
        hideFrontUI()
        hideBackUI()
        hideProcessButton()
        side = 0;
        
        //UI Initialization ends
        
        // Create and add the view to the screen.
        if(username == "" || password == "" || subscription == "" || url == ""){
            return;
        }
        DispatchQueue.main.async() {
            self.progressHUD = ProgressHUD(text: "Validating key")
            self.view.addSubview(self.progressHUD)
        }
        instance = AcuantMobileSDKController.initAcuantMobileSDK(withUsername: username, password:password, subscription:subscription, url:url, andDelegate: self)
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //UI
    func showFrontUI(){
        frontCardImageView.layer.cornerRadius = 8.0
        frontCardImageView.clipsToBounds = true
        frontCardImageView.isHidden=false;
        frontCardImageView.image = nil;
        frontCardImageView.layer.borderWidth=2
        frontImageViewLabel.isHidden=false
    }
    func hideFrontUI(){
        frontCardImageView.layer.cornerRadius = 8.0
        frontCardImageView.clipsToBounds = true
        frontCardImageView.isHidden=true;
        frontCardImageView.image = nil;
        frontCardImageView.layer.borderWidth=2
        frontImageViewLabel.isHidden=true
    }
    
    func showBackUI(){
        backCardImageView.layer.cornerRadius = 8.0
        backCardImageView.clipsToBounds = true
        backCardImageView.isHidden=false;
        backCardImageView.image = nil;
        backCardImageView.layer.borderWidth=2
        backImageViewLabel.isHidden=false;
    }
    
    func hideBackUI(){
        backCardImageView.layer.cornerRadius = 8.0
        backCardImageView.clipsToBounds = true
        backCardImageView.isHidden=true;
        backCardImageView.image = nil;
        backCardImageView.layer.borderWidth=2
        backImageViewLabel.isHidden=true;
    }
    
    
    func showProcessButton(){
        processButton.isHidden=false;
        processButton.isUserInteractionEnabled=true;
        processButton.isEnabled=true;
    }
    
    func hideProcessButton(){
        processButton.isHidden=true;
    }
    
    func resetData(){
        side = 0;
        resultData = nil;
        backCardImageView.image = nil;
        frontCardImageView.image = nil;
    }
    
    
    //Show Camera
    func showCamera(){
        if(instance != nil){
            instance.showManualCameraInterface(in: self, delegate: self, cardType: AcuantCardTypeDriversLicenseCard, region:AcuantCardRegionUnitedStates, andBackSide: true)
        }
    }
    
    
    // UI Callbacks
    func didFinishShowingResult() {
        resetData()
        hideProcessButton()
        hideFrontUI()
        hideBackUI()
    }
    
    
    func showBackButton() -> Bool {
        return true;
    }
    
    
    //User Actions
    @IBAction func captureTapped(_ sender: Any) {
        showFrontUI()
        hideBackUI()
        resetData()
    }
    
    @IBAction func processTapped(_ sender: Any) {
        let errorMessage : String = validateState()
        if(errorMessage == ""){
            DispatchQueue.main.async() {
                if(self.progressHUD == nil){
                    self.progressHUD = ProgressHUD(text: "Capturing data")
                    self.view.addSubview(self.progressHUD)
                    
                }else{
                    self.progressHUD.text = "Capturing data";
                    self.progressHUD.show()
                }
            }
            
            let options : AcuantCardProcessRequestOptions  = AcuantCardProcessRequestOptions.defaultRequestOptions(for: cardType);
            instance.processCard(with: options, frontImage: frontCardImageView.image, back: backCardImageView.image, barcodeString: nil)
            
        }else{
            let alertController = UIAlertController(title: "AcuantSwiftSample", message:
                errorMessage, preferredStyle: UIAlertController.Style.alert)
            alertController.addAction(UIAlertAction(title: "Dismiss", style: UIAlertAction.Style.default,handler: nil))
            
            self.present(alertController, animated: true, completion: nil)
        }

    }
    
    @IBAction func frontImageTapped(_ sender: Any) {
        side = 0;
        showCamera()
    }
    
    @IBAction func backImageTapped(_ sender: Any) {
        side = 1;
        showCamera()
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "toResult"{
            let destinationVC :ResultViewController = segue.destination as! ResultViewController
            destinationVC.cancelDelegate = self
            destinationVC.cardData=resultData as! NSDictionary
            destinationVC.cardType=cardType
            destinationVC.username=username
            destinationVC.password=password
        }
        
    }
    
    
    //Data validation
    func validateState()->String{
        var retValue : String!
        retValue = "";
        if(frontCardImageView.image==nil){
            retValue = "Please provide an ID image";
        }
        return retValue;
    }
    
    
    
    // Web service callbacks
    func mobileSDKWasValidated(_ wasValidated: Bool) {
        DispatchQueue.main.async() {
            self.progressHUD.hide()
        }
        if(wasValidated){
            print("valid license key")
        }else{
            let alertController = UIAlertController(title: "AcuantSwiftSample", message:
                "Credntials are not valid", preferredStyle: UIAlertController.Style.alert)
            alertController.addAction(UIAlertAction(title: "Dismiss", style: UIAlertAction.Style.default,handler: nil))
            
            self.present(alertController, animated: true, completion: nil)
        }
    }
    
    func didFinishProcessingCard(with result: AcuantCardResult!) {
        
        
    }
    
    func didFinishProcessingCard(withAssureIDResult json: Any!) {
        resultData=json;
        DispatchQueue.main.async() {
            self.progressHUD.hide()
            self.performSegue(withIdentifier: "toResult", sender: nil)
        }
    }
    
    // Capture Callbacks
    func didCaptureCropImage(_ cardImage: UIImage!, scanBackSide: Bool, andCardType cardType: AcuantCardType, withImageMetrics imageMetrics: [AnyHashable : Any]!) {
        self.cardType=cardType
        if(cardType == AcuantCardTypeDriversLicenseCard){
            if(side==0){
                frontCardImageView.image=cardImage
                let alertController = UIAlertController(title: "AcuantSwiftSample", message:
                    "Scan the backside of the drivers license", preferredStyle: UIAlertController.Style.alert)
                alertController.addAction(UIAlertAction(title: "OK", style: UIAlertAction.Style.default,handler:{(alert: UIAlertAction!) in
                    self.showCamera()
                    self.side=1
                }))
                self.present(alertController, animated: true, completion: nil)
            }else{
                showBackUI()
                backCardImageView.image=cardImage
                backImageViewLabel.isHidden=true
                showProcessButton();
            }
        }else{
            frontCardImageView.image=cardImage
            showProcessButton();
        }
        
    }
    
    func didCaptureData(_ data: String!) {
        
    }
    
    
    func didFailToCaptureCropImage() {
        DispatchQueue.main.async() {
            let alertController = UIAlertController(title: "AcuantSwiftSample", message:
                "Failed to capture Image.Please try again", preferredStyle: UIAlertController.Style.alert)
            alertController.addAction(UIAlertAction(title: "Dismiss", style: UIAlertAction.Style.default,handler: nil))
            
            self.present(alertController, animated: true, completion: nil)
        }
    }
    
    func didFinishValidatingImage(with result: AcuantCardResult!) {
        
    }
    
    
    // Errors
    func didFailWithError(_ error: AcuantError!) {
        if(progressHUD != nil){
            progressHUD.hide()
        }
        showErrorAlert(error)
        
    }
    
    func showErrorAlert(_ error: AcuantError!){
        DispatchQueue.main.async() {
            let alertController = UIAlertController(title: "AcuantSwiftSample", message:
                error.errorMessage, preferredStyle: UIAlertController.Style.alert)
            alertController.addAction(UIAlertAction(title: "Dismiss", style: UIAlertAction.Style.default,handler: nil))
            
            self.present(alertController, animated: true, completion: nil)
        }
    }
    
}

